ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,72 @@
+
+The approach to make this pass/fail detection possible proposed here, is to use historical results as a benchmark – not only actual results of the performance test, but also monitoring data about the resource usage, etc.
+
+> Example: Imagine you develop a web e-shop application that uses PostgreSQL data backend. Your perf test is browsing through a goods catalog and measures latency of pages. When you want to decide on pass or fail result of this test, you need to check that metrics like below are aligned to previous results of a same test with same configuration:
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Checking minimum and maximum of the metric in historical data and making sure the new result is in this range.
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some machine learning thingy.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in Stonesoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
+
+We are considering two different options here. They differ in what portion of logic mentioned above is executed where.
+
+### Glossary
+
+- SUT: Subject Under Test
+- Horreum: a results storage service: <https://github.com/Hyperfoil/Horreum>
+- CPT: Continuous Performance Testing
+
+## Decision
+
+Let's use this architecture with a single Horreum instance that would be managed by Red Hat and used by all customers.
+
+![Architecture diagram with Horreum](assets/0018-apps-continuous-perf-testing.svg "Architecture diagram with Horreum")
+
+1. Performance test runs in the Tekton pipeline and generates JSON with test parameters and results.
```

Comment:
Are there standards for this JSON format? Do we need to be prescriptive of how to record data? Have we considered what would be required for tenant isolation?

Reply:
> Are there standards for this JSON format?

It can be any valid JSON - Horreum can handle it. I would consider it implementation detail as of now.

> Do we need to be prescriptive of how to record data?

No, because it will be different for every test - if you are testing an web app, you will be interested in latency and number number of requests per second and failure rate... and if you are testing application processing kafka messages, you will measure per message processing duration and amount of data processed...

> Have we considered what would be required for tenant isolation?

If you are asking about isolating test runs itself: Yes, noisy neighbours will cause performance test results will vary and it is up to test design, deployment requirements (e.g. make sure your pods have requests and limits configured) and results evaluation to overcome this (we have some experience with it).

If you are asking about if Horreum can handle multiple tenants: Yes, Horreum uses Keycloak for fine grained user permissions.

Reply:
Fine to leave it as an implementation detail for now, I just wanted to make sure that it was one and that we didn't need to identify some standard.

Re: multiple tenants, I was mostly interested in ensuring that tests in one workspace wouldn't be able to populate/retrieve data from another workspace. This sounds like Keycloak would be sufficient. Can you just call this bit out somewhere as a reminder if its importance?

Reply:
Thank you! Reported https://github.com/Hyperfoil/Horreum/issues/420 and added a line to the ADR about it to "Cons" section.